### ISSUE TITLE
[clang: static analyzer] False positive of clang-analyzer-core.uninitialized.Assign with std::addressof

### ISSUE THREAD

**Issue by igkper** [llvm/llvm-project#59936]
When an uninitialized object has its value set via the result of `std::addressof` either via assignment or placement new, the static analyzer does not recognize that the value is no longer uninitialized since it issues `clang-analyzer-core.uninitialized.Assign` when that object is subsequently used to construct/assign another. See the example below (also shown at https://godbolt.org/z/GMjr4sv9c)

```cpp
#include <memory>

void rejected()
{
    int x;
    *(std::addressof(x)) = 1;
    int y = x;  // warning: Assigned value is garbage or undefined [clang-analyzer-core.uninitialized.Assign]
}

void accepted()
{
    int x;
    *(&x) = 1;
    int y = x;
}
```

**Comment by haoNoQ** [comment]
Ok so a little birdie told me that this used to work after `LValueBitCast` was modeled correctly circa 2013. These days it resolves to `__builtin_addressof()` instead, but we have an explicit model for that as well.

However, it looks like `std::addressof()` calls are no longer looked into during analysis ("inlined") so the model has no chance to kick in. Must be one of the standard library inlining heuristics acting up. Separately, for some reason the by-reference argument doesn't get invalidated during "conservative evaluation" of `std::addressof()`, which appears to be a bug in and of itself.

Thanks for the useful bug report! Sounds like a couple easy fixes should get this right.

**Comment by igkper** [comment]
Sounds great, thanks!

Den ons 11 jan. 2023 kl 22:23 skrev NoQ ***@***.***>:

**Comment by igkper** [comment]
Sounds great, thanks!

**Comment by igkper** [comment]
Sounds great, thanks!

**Comment by bartdesmet** [comment]
+1 - We ran into this as well for code like `std::construct_at(std::addressof(p), args)`.

**Comment by bernardosulzbach** [comment]
Commenting here because it looks like I am running into another manifestation of the same underlying problem.

```cpp
#include <utility>

void f(int *write);

int readX() {
  int x;
  f(std::addressof(x));
  return x;
}
```

produces

> warning: Undefined or garbage value returned to caller [clang-analyzer-core.uninitialized.UndefReturn]

with clang-tidy 18.1.6.
